akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedRecoveringSnapshot.scala

```diff
@@ -0,0 +1,202 @@
+      override def snapshotPluginId: String = b.snapshotPluginId
+    }
+  }
+
+  /**
+   * Called whenever a message replay fails. By default it logs the error.
+   *
+   * The actor is always stopped after this method has been invoked.
+   *
+   * @param cause failure cause.
+   * @param event the event that was processed in `receiveRecover`, if the exception was thrown there
+   */
+  protected def onRecoveryFailure(cause: Throwable, event: Option[Any]): Behavior[Any] = {
+    cancelRecoveryTimer()
+    event match {
+      case Some(evt) ⇒
+        log.error(cause, "Exception in receiveRecover when replaying event type [{}] with sequence number [{}] for " +
+          "persistenceId [{}].", evt.getClass.getName, lastSequenceNr, persistenceId)
+        Behaviors.stopped
+
+      case None ⇒
+        log.error(cause, "Persistence failure when replaying events for persistenceId [{}]. " +
+          "Last known sequence number [{}]", persistenceId, lastSequenceNr)
+        Behaviors.stopped
+    }
+  }
+
+  // FIXME separate the waiting for snapshot state from the waiting for events one // EventsourcedSnapshotRecovery -> EventsourcedRecovery
+  protected def onRecoveryTick(snapshot: Boolean): Behavior[Any] =
+    // we know we're in snapshotting mode
+    if (snapshot) onRecoveryFailure(new RecoveryTimedOut(s"Recovery timed out, didn't get snapshot within $timeout"), event = None)
+    else same // ignore, since we received the snapshot already
+
+  // ----------
+
+  override def onMessage(msg: Any): Behavior[Any] = {
+    log.info(s"[${Logging.simpleName(getClass)}] onMessage: " + msg)
+    msg match {
+      // TODO explore crazy hashcode hack to make this match quicker...?
+      case SnapshotterResponse(r)      ⇒ onSnapshotterResponse(r)
+      case JournalResponse(r)          ⇒ onJournalResponse(r)
+      case RecoveryTickEvent(snapshot) ⇒ onRecoveryTick(snapshot = snapshot)
+      case c: Command @unchecked       ⇒ onCommand(c.asInstanceOf[Command]) // explicit cast to fail eagerly
+    }
+  }
+
+  // ----------
+
+  // ---------- journal interactions ---------
+
+  /**
+   * Instructs the snapshot store to load the specified snapshot and send it via an [[SnapshotOffer]]
+   * to the running [[PersistentActor]].
+   */
+  private def loadSnapshot(persistenceId: String, criteria: SnapshotSelectionCriteria, toSequenceNr: Long): Unit = {
+    snapshotStore.tell(LoadSnapshot(persistenceId, criteria, toSequenceNr), selfUntypedAdapted)
+  }
+
+  private def returnRecoveryPermitOnlyOnFailure(cause: Throwable): Unit = {
+    log.debug("Returning recovery permit, on failure because: " + cause.getMessage)
+    // IMPORTANT to use selfUntyped, and not an adapter, since recovery permitter watches/unwatches those refs (and adapters are new refs)
+    extension.recoveryPermitter.tell(RecoveryPermitter.ReturnRecoveryPermit, selfUntyped)
+  }
+
+  override def toString = s"EventsourcedRecoveringSnapshot($persistenceId)"
+
+}
```

Comment:
I really like how these different states are different behaviors, and we will be able to change them :)
(the EventSourced trait is a nightmare)